cwl/src/main/scala/cwl/preprocessor/CwlPreProcessor.scala

```diff
@@ -0,0 +1,272 @@
+    Log.info(s"Salading ${file.pathAsString}")
+    CwlDecoder.saladCwlFile(file) 
+  }
+
+  implicit class PrintableJson(val json: Json) extends AnyVal {
+    def printCompact = io.circe.Printer.noSpaces.pretty(json)
+  }
+
+  private [preprocessor] implicit class EnhancedCwlId(val id: String) extends AnyVal {
+    def asReference: Option[CwlReference] = CwlReference.fromString(id)
+    def stripFilePrefix = id.stripPrefix(LocalScheme)
+  }
+}
+
+/**
+  * Class to create a standalone version of a CWL file.
+  * @param saladFunction function that takes a file and produce a saladed version of the content
+  */
+class CwlPreProcessor(saladFunction: BFile => Parse[String] = saladCwlFile) {
+
+  /**
+    * Pre-process a CWL file and create a standalone, runnable (given proper inputs), inlined version of its content.
+    *
+    * The general idea is to work on CwlReferences, starting from the one coming to this function in the form of file and optional root.
+    * The goal is to look at the steps in this workflow that point to other references, and recursively flatten them until we can replace the step with
+    * its flat version.
+    *
+    * There are 3 pieces of information that are carried around during this process:
+    *  1) ProcessedReferences: A Map[CwlReference, Json] of CwlReference for which we have the fully processed (saladed AND flattened) Json value.
+    *
+    *  2) UnProcessedReferences: A Map[CwlReference, Json] of CwlReference for which we have the saladed but NOT flattened Json value.
+    *     This can happen because a file can contain multiple tools / workflows. When we salad / parse this file, we get (CwlReference, Json) pairs
+    *     for all the workflow / tools in the file, but they are not flattened yet.
+    *     We keep this to avoid having to re-salad / re-parse files unnecessarily.
+    *
+    *  3) BreadCrumb: A List[CwlReference] used to follow the trail of CwlReferences that we are processing as we recurse down.
+    *     This is used to be able to detect circular dependencies (if the cwlReference being processed is in that list, then we have a circular dependency) .
+    *
+    */
+  def preProcessCwlFile(file: BFile, cwlRoot: Option[String]): Parse[Json] = {
+    val reference = CwlReference(file, cwlRoot)
+    def flatten: Parse[ProcessedJsonAndDependencies] = flattenCwlReference(CwlReference(file, cwlRoot), Map.empty, Map.empty, Set.empty)
+    
+    for {
+      original <- parseYaml(file.contentAsString)
+      flattened <- original
+        .asObject
+        .flatMap(_.kleisli("id").flatMap(_.asString).flatMap(CwlReference.fromString)) match {
+        // This by passes the pre-processing if the file already has an id for which the file doesn't match the path of the file
+        // passed to this function, as this would indicate that it has already been saladed and pre-processed.
+        case Some(contentRef) if !contentRef.file.equals(reference.file) => original.validParse
```

Comment:
This is not great and kinda hacky but the best way I could find to not pre-process files that have (very likely) already been processed.
I tried to make the whole thing idempotent to avoid this issue but turns out salading is not full idempotent and this code would also need a few adjustments to be idempotent. I'll try to give it another shot.